### DISCUSSION TITLE
Right way to test Serverless Plugins?

### DISCUSSION THREAD

**Discussion by ubaniabalogun** [serverless/serverless#9614] [Q&A] [answered]
Hi there, what's the right way to test a serverless plugin? 
So far, I've written a plugin and have unit-tests for my business logic. However, I'm unsure how to write tests to ensure that my plugin behaves correctly when called as part of Serverless frameworks lifecycle. 

I stumbled on the [runServerless util](https://github.com/serverless/test/blob/master/docs/run-serverless.md#run-serverless). Should framework plugins be using this utility as well? Or is it purely for testing Serverless framework internals? If it is purely internal, is there an alternative that plugin developers should use
Thanks,

**Comment by medikoo** [comment, marked as answer]
> I stumbled on the runServerless util. Should framework plugins be using this utility as well? Or is it purely for testing Serverless framework internals? If it is purely internal, is there an alternative that plugin developers should use
Thanks,

I'd highly recommend using `runServerless` util to test plugins, as it tests against full cycle, and it's the only reliable way to confirm all works as expected.

I believe it can be relatively easily used to test plugins, but please ask if you stumble on any issues

**Reply by ubaniabalogun** [reply]
Thanks for the information @medikoo . Just took a swing at setting this up in my project using the [README instructions here](https://github.com/serverless/test#readme) and I'm running into the below error when I execute `npm run test`

```
sls-apb ubalogun$ npm run test

> sls-apb@1.0.0 test /Users/ubalogun/Projects/socless_repos/sls-apb
> tsc && ./node_modules/.bin/mocha


✖ ERROR: Error: Cannot find module '@serverless/test/setup/async-leaks-detector'
Require stack:
- /Users/ubalogun/Projects/socless_repos/sls-apb/node_modules/mocha/lib/esm-utils.js
- /Users/ubalogun/Projects/socless_repos/sls-apb/node_modules/mocha/lib/mocha.js
- /Users/ubalogun/Projects/socless_repos/sls-apb/node_modules/mocha/lib/cli/one-and-dones.js
- /Users/ubalogun/Projects/socless_repos/sls-apb/node_modules/mocha/lib/cli/options.js
- /Users/ubalogun/Projects/socless_repos/sls-apb/node_modules/mocha/bin/mocha
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:625:15)
    at Function.Module._load (internal/modules/cjs/loader.js:527:27)
    at Module.require (internal/modules/cjs/loader.js:683:19)
    at require (internal/modules/cjs/helpers.js:16:16)
    at exports.requireOrImport (/Users/ubalogun/Projects/socless_repos/sls-apb/node_modules/mocha/lib/esm-utils.js:42:12)
    at exports.handleRequires (/Users/ubalogun/Projects/socless_repos/sls-apb/node_modules/mocha/lib/cli/run-helpers.js:94:34)
    at async /Users/ubalogun/Projects/socless_repos/sls-apb/node_modules/mocha/lib/cli/run.js:341:25 {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/Users/ubalogun/Projects/socless_repos/sls-apb/node_modules/mocha/lib/esm-utils.js',
    '/Users/ubalogun/Projects/socless_repos/sls-apb/node_modules/mocha/lib/mocha.js',
    '/Users/ubalogun/Projects/socless_repos/sls-apb/node_modules/mocha/lib/cli/one-and-dones.js',
    '/Users/ubalogun/Projects/socless_repos/sls-apb/node_modules/mocha/lib/cli/options.js',
    '/Users/ubalogun/Projects/socless_repos/sls-apb/node_modules/mocha/bin/mocha'
  ]
}
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! sls-apb@1.0.0 test: `tsc && ./node_modules/.bin/mocha`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the sls-apb@1.0.0 test script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/ubalogun/.npm/_logs/2021-06-20T05_20_53_742Z-debug.log
```


I've also searched through the `node_modules` directory for `async-leaks-detector` and don't find any references to it in the @serverless/test package. Are there additional steps to complete this setup that aren't documented in the readme?

Thanks,

**Reply by medikoo** [reply]
@ubaniabalogun ensure to install `@serverless/test` in your project

**Reply by medikoo** [reply]
Sorry, wrong answer. Documentation was not updated after `async-leaks-detector` hook was removed. I've just updated documentation, to provide an up to date setup version.

**Reply by ubaniabalogun** [reply]
Thanks for the update @medikoo. I'm hitting a different bug now, which I take as a sign of progress 😄 

Here's the new bug I'm hitting
```
Provided 'serverlessPath' (/private/var/folders/t6/dwgv3yfs25x69rd9nvrfplrh0000gn/T/tmpdirs-serverless/9436/838673/test/mock-service) doesn't point a working node module: Cannot find module '/private/var/folders/t6/dwgv3yfs25x69rd9nvrfplrh0000gn/T/tmpdirs-serverless/9436/838673/test/mock-service'
Require stack:
- /Users/ubalogun/Projects/socless_repos/sls-apb/node_modules/@serverless/test/run-serverless.js
- /Users/ubalogun/Projects/socless_repos/sls-apb/test/test_slsApb.js
- /Users/ubalogun/Projects/socless_repos/sls-apb/node_modules/mocha/lib/esm-utils.js
- /Users/ubalogun/Projects/socless_repos/sls-apb/node_modules/mocha/lib/mocha.js
- /Users/ubalogun/Projects/socless_repos/sls-apb/node_modules/mocha/lib/cli/one-and-dones.js
- /Users/ubalogun/Projects/socless_repos/sls-apb/node_modules/mocha/lib/cli/options.js
- /Users/ubalogun/Projects/socless_repos/sls-apb/node_modules/mocha/bin/mocha
```

I think part of why I'm hitting this bug is that i'm unclear what the serverlessDir value should be set to for runServerless. 
Any chance you can give me a minimal example of `runServerless` being used in a simple 3rd party plugin please?

**Reply by medikoo** [reply]
@ubaniabalogun please check the docs: https://github.com/serverless/test/blob/master/docs/run-serverless.md#usage

_`serverlessDir` should point a path to Serverless Framework directory which we want to run test against._

You can also inspect a bit more advanced `runServerless` setup (with fixtures engine) which is done in context of dashboard plugin -> https://github.com/serverless/dashboard-plugin/blob/2c8c599f815b2d12b3323f9f23c2f34f3b267e7b/test/run-serverless.js

**Reply by ubaniabalogun** [reply]
Thanks so very much for your responses @medikoo! I was finally able to get it working!

**Reply by jiw-mh** [reply]
@medikoo It seems serverless/test is gone. What is the new process?

**Reply by medikoo** [reply]
@jiw-mh I'm no longer involved in this project. By looking at recent contributions I think @eahefnawy or @czubocha may have the answers.
